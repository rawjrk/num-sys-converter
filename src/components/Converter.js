import convert from "../lib/convert";
import { useCachedState } from "../lib/hooks";
import { Form, Label, Input, Select, Options } from "./Form";
import { optionsPropType } from "./Form/Options";

function Converter({ options }) {
  const [num, setNum] = useCachedState("number", "");
  const [fromBase, setFromBase] = useCachedState("frombase", 10);
  const [toBase, setToBase] = useCachedState("tobase", 2);
  const result = num ? convert(num, fromBase, toBase) : "";

  return (
    <Form id="converter">
      <Label for="number">Number</Label>
      <Input
        id="number"
        type="text"
        placeholder="Input here..."
        value={num}
        onChange={(e) => setNum(e.target.value)}
      />

      <Label for="frombase">From</Label>
      <Select
        id="frombase"
        aria-label="Select number base to convert from"
        defaultValue={fromBase}
        onChange={(e) => setFromBase(e.target.value)}
      >
        <Options options={options} />
      </Select>

      <Label for="tobase">To</Label>
      <Select
        id="tobase"
        aria-label="Select number base to convert to"
        defaultValue={toBase}
        onChange={(e) => setToBase(e.target.value)}
      >
        <Options options={options} />
      </Select>

      <Label for="result">Result</Label>
      <Input
        id="result"
        type="text"
        disabled
        value={result}
        onChange={(e) => setNum(e.target.value)}
      />
    </Form>
  );
}

Converter.propTypes = {
  options: optionsPropType,
};

Converter.defaultProps = {
  options: [],
};

export default Converter;
