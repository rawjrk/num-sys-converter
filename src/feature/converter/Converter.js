import { useCachedState } from "../../lib/hooks";
import Form from "./components/Form";
import Label from "./components/Label";
import Input from "./components/Input";
import Select from "./components/Select";
import Options, { optionsPropType } from "./components/Options";
import convert from "./convert";

function Converter({ options }) {
  const [num, setNum] = useCachedState("number", "");
  const [fromBase, setFromBase] = useCachedState("frombase", 10);
  const [toBase, setToBase] = useCachedState("tobase", 2);

  return (
    <Form id="converter">
      <Label for="number">Number</Label>
      <Input
        id="number"
        type="text"
        placeholder="Input here..."
        value={num}
        onChange={(e) => setNum(e.target.value)}
      />

      <Label for="frombase">From</Label>
      <Select
        id="frombase"
        aria-label="Select number base to convert from"
        defaultValue={fromBase}
        onChange={(e) => setFromBase(e.target.value)}
      >
        <Options options={options} />
      </Select>

      <Label for="tobase">To</Label>
      <Select
        id="tobase"
        aria-label="Select number base to convert to"
        defaultValue={toBase}
        onChange={(e) => setToBase(e.target.value)}
      >
        <Options options={options} />
      </Select>

      <Label for="result">Result</Label>
      <Input
        id="result"
        type="text"
        disabled
        value={num ? convert(num, fromBase, toBase) : ""}
      />
    </Form>
  );
}

Converter.propTypes = {
  options: optionsPropType,
};

Converter.defaultProps = {
  options: [],
};

export default Converter;
