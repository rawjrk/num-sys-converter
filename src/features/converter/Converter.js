import { useSelector, useDispatch } from "react-redux";
import { Form, Label, Input, Select, Options } from "../shared";
import {
  setNumber,
  setFromBase,
  setToBase,
  selectConverter,
} from "./converterSlice";

import options from "./options.json";
import convert from "./convert";

function Converter() {
  const { number, fromBase, toBase } = useSelector(selectConverter);
  const dispatch = useDispatch();

  const onNumberChange = (e) => dispatch(setNumber(e.target.value));
  const onFromBaseChange = (e) => dispatch(setFromBase(Number(e.target.value)));
  const onToBaseChange = (e) => dispatch(setToBase(Number(e.target.value)));
  const result = number ? convert(number, fromBase, toBase) : "";

  return (
    <Form id="converter">
      <Label htmlFor="number">Number</Label>
      <Input
        id="number"
        type="text"
        placeholder="Input here..."
        value={number}
        onChange={onNumberChange}
      />

      <Label htmlFor="frombase">From</Label>
      <Select
        id="frombase"
        aria-label="Select number base to convert from"
        defaultValue={fromBase}
        onChange={onFromBaseChange}
      >
        <Options options={options} />
      </Select>

      <Label htmlFor="tobase">To</Label>
      <Select
        id="tobase"
        aria-label="Select number base to convert to"
        defaultValue={toBase}
        onChange={onToBaseChange}
      >
        <Options options={options} />
      </Select>

      <Label htmlFor="result">Result</Label>
      <Input id="result" type="text" disabled value={result} />
    </Form>
  );
}

export default Converter;
