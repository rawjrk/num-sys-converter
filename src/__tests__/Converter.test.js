import { screen, fireEvent } from "@testing-library/react";
import Converter from "../features/converter/Converter";

const setupOptions = {
  includeTheme: true,
  preloadedState: {
    converter: { number: "13", fromBase: 10, toBase: 2 },
  },
};

const setup = () => {
  const render = renderWithProviders(<Converter />, setupOptions);

  const numberInput = screen.getByLabelText("Number");
  const fromSelect = screen.getByLabelText("From");
  const toSelect = screen.getByLabelText("To");
  const resultBox = screen.getByLabelText("Result");

  return { ...render, numberInput, fromSelect, toSelect, resultBox };
};

test("all labels in place", () => {
  setup();

  expect(screen.getByText(/Number/i)).toBeInTheDocument();
  expect(screen.getByText(/From/i)).toBeInTheDocument();
  expect(screen.getByText(/To/i)).toBeInTheDocument();
  expect(screen.getByText(/Result/i)).toBeInTheDocument();
});

describe("storage state and form inputs", () => {
  test("match initial values", () => {
    const { store, numberInput, fromSelect, toSelect, resultBox } = setup();
    const { converter } = store.getState();

    expect(numberInput).toHaveValue("13");
    expect(converter.number).toBe("13");
    expect(fromSelect).toHaveDisplayValue(/decimal/i);
    expect(converter.fromBase).toBe(10);
    expect(toSelect).toHaveDisplayValue(/binary/i);
    expect(converter.toBase).toBe(2);
    expect(resultBox).toHaveValue("1101");
  });

  test("amend on change", () => {
    const { store, numberInput, fromSelect, toSelect, resultBox } = setup();

    fireEvent.change(numberInput, { target: { value: "57" } });
    fireEvent.change(fromSelect, { target: { value: "8" } });
    fireEvent.change(toSelect, { target: { value: "16" } });

    const { converter } = store.getState();

    expect(numberInput).toHaveValue("57");
    expect(converter.number).toBe("57");
    expect(fromSelect).toHaveDisplayValue(/octal/i);
    expect(converter.fromBase).toBe(8);
    expect(toSelect).toHaveDisplayValue(/hex/i);
    expect(converter.toBase).toBe(16);
    expect(resultBox).toHaveValue("2f");
  });
});

test("matches snapshot", () => {
  const tree = snapshotWithProviders(<Converter />, setupOptions);
  expect(tree).toMatchSnapshot();
});
