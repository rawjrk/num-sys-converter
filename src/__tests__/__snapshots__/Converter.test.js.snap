// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className="Form-sc-1ufwmkm-0 etyVqS"
  id="converter"
>
  <label
    className="Label-sc-p42w1g-0 jPAJlC"
    htmlFor="number"
  >
    Number
  </label>
  <input
    className="Input-sc-1qsdt6o-0 gKPVqn"
    id="number"
    onChange={[Function]}
    placeholder="Input here..."
    type="text"
    value="13"
  />
  <label
    className="Label-sc-p42w1g-0 jPAJlC"
    htmlFor="frombase"
  >
    From
  </label>
  <select
    aria-label="Select number base to convert from"
    className="Select-sc-31pzb-0 cuqxij"
    defaultValue={10}
    id="frombase"
    onChange={[Function]}
  >
    <option
      id={2}
      value={2}
    >
      Binary [base-2]
    </option>
    <option
      id={8}
      value={8}
    >
      Octal [base-8]
    </option>
    <option
      id={10}
      value={10}
    >
      Decimal [base-10]
    </option>
    <option
      id={16}
      value={16}
    >
      Hexadecimal [base-16]
    </option>
  </select>
  <label
    className="Label-sc-p42w1g-0 jPAJlC"
    htmlFor="tobase"
  >
    To
  </label>
  <select
    aria-label="Select number base to convert to"
    className="Select-sc-31pzb-0 cuqxij"
    defaultValue={2}
    id="tobase"
    onChange={[Function]}
  >
    <option
      id={2}
      value={2}
    >
      Binary [base-2]
    </option>
    <option
      id={8}
      value={8}
    >
      Octal [base-8]
    </option>
    <option
      id={10}
      value={10}
    >
      Decimal [base-10]
    </option>
    <option
      id={16}
      value={16}
    >
      Hexadecimal [base-16]
    </option>
  </select>
  <label
    className="Label-sc-p42w1g-0 jPAJlC"
    htmlFor="result"
  >
    Result
  </label>
  <input
    className="Input-sc-1qsdt6o-0 gKPVqn"
    disabled={true}
    id="result"
    type="text"
    value="1101"
  />
</form>
`;
